galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,300 @@
+			scope.Debugf("Received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// Delete the Service and ServiceEntry
+		delete(p.services, fullName)
+		delete(p.mcpResources, fullName)
+		p.updateVersion()
+	default:
+		scope.Errorf("Unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) handleEndpointsEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the version has changed.
+		curEntry, ok := p.endpoints[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("Received endpoints for the current, known version: %v", event)
+			return
+		}
+
+		p.endpoints[fullName] = entry
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
+		if !ok {
+			// TODO:
+			return
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(svcEntry, entry.Item.(*coreV1.Endpoints))
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		delete(p.endpoints, fullName)
+		p.updateVersion()
+	default:
+		scope.Errorf("Unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) setMcpEntry(fullName resource.FullName, mcpEntry *mcp.Resource) {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	p.mcpResources[fullName] = mcpEntry
+}
+
+func (p *Handler) deleteMcpResource(fullName resource.FullName) {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	delete(p.mcpResources, fullName)
+}
+
+func (p *Handler) updateVersion() {
+	p.version++
+	monitoring.RecordStateTypeCount(collection.String(), len(p.mcpResources))
+
+	scope.Debugf("In-memory state has changed:\n%v\n", p)
+	p.pendingEvents++
+	if p.distribute {
+		p.notifyChanged()
+	}
+}
+
+func (p *Handler) notifyChanged() {
+	p.listener.CollectionChanged(collection)
+}
+
+func (p *Handler) versionString() string {
+	return fmt.Sprintf("%d", p.version)
+}
+
+func (p *Handler) toMcpResource(svcEntry resource.Entry, endpoints *coreV1.Endpoints) (*mcp.Resource, bool) {
+	fullName := svcEntry.ID.FullName
+	svc := svcEntry.Item.(*coreV1.ServiceSpec)
+	svcMeta := svcEntry.Metadata
+	se := convert.Service(svc, svcMeta, svcEntry.ID, p.domainSuffix)
+
+	// Set the endpoints, if available.
+	if endpoints != nil {
+		se.Endpoints = convert.Endpoints(svcMeta, endpoints, p.localityStrategy)
```

Comment:
This is an overwrite.

Reply:
yes ... the endpoints object should have all of the latest.